### DISCUSSION TITLE
Use of Complex Eigen Matrices / general linear operator

### DISCUSSION THREAD

**Discussion by insyzygy** [autodiff/autodiff#212] [Q&A]
I had two related questions. 

1. I see the example for complex scalars in forward mode. Is there a way to use complex Eigen Vector / Matrix objects (e.g. MatrixXcd) in forward or reverse mode to compute gradients. I had trouble getting this to work.
2. If I have a linear operator implemented as a function as opposed to a concrete matrix (where I know the analytic derivative since its linear), is there a way to wrap that as a functor (with maybe some other methods) so that autodiff can differentiate through it. The specific case I am interested in is a fast fourier transform (FFT) where it would be great if it was possible to wrap up the fft call with some other code so that autodiff could differentiate through an expression involving FFTs. In this case the function would be operating on complex numbers as well so its related to 1. 

Thanks!

**Comment by allanleal** [comment]
Hi @anamorphistic . Could you please post a simple example of what you tried? This would help me a lot. Thanks.

**Reply by insyzygy** [reply]
Here is an example showing where I have gotten and some questions in the comments with a mocked up idea of what I was hoping to do.  I actually would like a real valued derivative of a real valued expression with complex operations as intermediate steps. Not sure how to get that to work.  This example will run but only takes a complex derivative of a complex expression. Note I had to explicitly specialize 
```cpp
 NumTraits<autodiff::Real<N, std::complex<T> >> : NumTraits<double> 
```
to get conjugation to work in Eigen expressions, not sure if that is the right approach.


I tried to include a sketch in the comments of what I meant by wrapping a linear operator.  Curious if something like that could be made to work. 

```cpp// C++ includes
#include <iostream>
// autodiff include
#include <autodiff/forward/real.hpp>
#include <autodiff/forward/real/eigen.hpp>

using namespace autodiff;

namespace autodiff::detail {

    template<typename T>
    struct ArithmeticTraits<std::complex<T>> : ArithmeticTraits<T> {};

    //Need to define conjugate for complex expressions
    //Need to think about it but this seems to give a Wirtinger derivative
    //for derivative of a complex expression with respect to a complex variable
    // https://en.wikipedia.org/wiki/Wirtinger_derivatives
    template<size_t N, typename T>
    constexpr auto conj(const autodiff::detail::Real<N, std::complex<T>>& x)
    {

        autodiff::detail::Real<N, std::complex<T>> xconj;
        xconj[0] = std::conj(x[0]);
        autodiff::detail::For<1, N + 1>([&](auto i) constexpr {
            xconj[i] =0;
            });
        return xconj;
    }

}

typedef Real<1,std::complex<double> > creal;
typedef Eigen::Matrix<creal, -1, 1, 0> VectorXcmpx;
typedef Eigen::Matrix<creal, -1, -1, 0> MatrixXcmpx;

namespace Eigen {

    template<typename T>
    struct NumTraits;

    //Needed to do this to set IsComplex =1 so conjugate does something in Eigen expressions
    template<size_t N, typename T>
    struct NumTraits<autodiff::Real<N, std::complex<T> >> : NumTraits<double> // permits to get the epsilon, dummy_precision, lowest, highest functions
    {
        typedef autodiff::Real<N, T> Real; //worried this should be Real<N,std::complex<T>>> Real but that doesn't compile
        typedef autodiff::Real<N, T> NonInteger;
        typedef autodiff::Real<N, T> Nested;
        enum
        {
            IsComplex = 1,
            IsInteger = 0,
            IsSigned = 1,
            RequireInitialization = 1,
            ReadCost = 1,
            AddCost = 3,
            MulCost = 3
        };
    };
}

//simple quadratic form with complex matrix
creal f_cmpx(const VectorXcmpx& x,const Eigen::MatrixXcd &m)
{
    return x.adjoint()*m*x;
}

using namespace std::complex_literals;

/*  Question 1: 
// I would like to be able to have a real valued function with real valued vector inputs 
// that involves complex matrix operations as intermediate steps for example as below
// where the real inputs are phases of a complex phasor.
 
real f_phase(const VectorXreal& x,const Eigen::MatrixXcd &m)
{   
    VectorXCmpx z(x.rows());
    for(int ii=0; ii<x.rows();ii++){
        z[ii]=exp(1i*x[ii]);
    }
    return (z.adjoint()*m*z).abs();
}

*/


/*Question 2: 

I would like to be able to  define a representation of a linear operator in a class for example something like what is shown below. 
This could wrap a call to something like FFTW which is more efficient than explicitly representating it as a matrix. 

struct MyLinear {
    ...

    Eigen::MatrixXcd matrix_matrix_product(const Eigen::MatrixXcd &m){
        ... efficient code for linear operator acting on m e.g. call to FFTW
    }

    Eigen::VectorXcd matrix_vector_product(const Eigen::VectorXcd &v){
        ... efficient code for linear operator acting on v e.g. call to FFTW
    }
    ... whatever else is needed 
}


//wrap input T object to allow it to work in expressions we want to differentiate
template<class T> 
VectorXcmpx apply_operator(T operator_struct, VectorXcmpx& v){
    // use matrix matrix and matrix vector product to propagate values and derivatives using
    // fact that operator_struct represents a linear operator
    ... ?
}

// Use linear operator object in expression to differentiate through
real f_phase_FFT(const VectorXreal& x,const Eigen::MatrixXcd &m)
{   

    MyLinear linear_operator;
    VectorXCmpx z(x.rows());
    for(int ii=0; ii<x.rows();ii++){
        z[ii]=exp(1i*x[ii]);
    }
    return ( z.adjoint()*apply_operator(linear_operator,z)).abs();
}

*/


int main()
{
    Eigen::MatrixXcd c(3,3);
    VectorXcmpx z(3);
    z<<1.0+1i,2.0+2i,3.0+3i;
    c<<1,1i,2i,-1i,2,3i,-2i,-3i,3;
    
    creal u;                                     
    Eigen::VectorXcd g;
    
    g = gradient(f_cmpx, wrt(z), at(z,c), u); // evaluate the function value u and its gradient vector g = du/dx
    std::cout << "u = " << u << std::endl;    // print the evaluated output u
    std::cout << "g = \n" << g << std::endl;  // print the evaluated gradient vector g = du/dx
    
}

**Reply by allanleal** [reply]
Hi @anamorphistic - Have you tried 

`std::complex<autodiff::Real<N, T>>` 

instead of 

`autodiff::Real<N, std::complex<T>>`? 

With the first one, both the real and imaginary parts are `real` numbers, and you should be able to take derivatives with respect to them individually. My line of research doesn't need complex numbers, so this is something I haven't explored much in `autodiff`, but if you find a good solution for this, I invite you to create a PR with the necessary changes and some examples for the website. 

This would be a fantastic addition to the project as it seems like there are a lot of people interested in more support for complex numbers (which unfortunately I can't work on right now).

**Reply by insyzygy** [reply]
Is it correct that using ```std::complex<autodiff::Real<N, T>>``` would require basically creating a ```complex.hpp``` that would be overloads of most of what is in ```forward/real/real.hpp``` for ```std::complex<autodiff::Real<N, T>>``` arguments.  Is that the idea? I'll think about that. 

I was also curious, do you have any suggestions on how to do this (support complex numbers) for the reverse mode case, or where to start. I abstractly understand what the code is doing, but at the moment I don't really have a handle on exactly how  the expressions templates all fit together and what would need to be changed to make that work. 

Thanks.
